### PR TITLE
Collection View data source creator

### PR DESCRIPTION
**Pull request by MartinJibber**
I deprecated the Collection View manager and started splitting up its responsibilities into more lightweight pieces.

To start, I created a CollectionViewDataSourceCreator protocol. More details are annotated.

### CODE REVIEW

File: Shared/Extensions/Task+Extensions.swift [hunk 1]
```diff
@@ -29,18 +29,22 @@ extension Task where Success == Void, Failure == Never {
 
 extension Task where Success == Never, Failure == Never {
 
+    /// Suspends the current task for _at least_ the given duration
+    /// in seconds.
+    static func sleep(seconds: TimeInterval) async {
+        await Task.sleep(UInt64(seconds * 1_000_000_000))
+    }
```
Comment (MartinJibber) [inline Shared/Extensions/Task+Extensions.swift:36]: Just a convenience so you don't have to right the time in nanoseconds.

File: Shared/Extensions/UI Extensions/UICollectionView+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+//  UICollectionView+Extensions.swift
+//  UICollectionView+Extensions
+//
+//  Created by Martin Young on 8/24/21.
+//  Copyright © 2021 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+extension UICollectionView {
```
Comment (MartinJibber) [inline Shared/Extensions/UI Extensions/UICollectionView+Extensions.swift:11]: I grabbed these from the old CollectionViewManager and made them async. We're dumbing down the collection view manager so it doesn't need these any more.

File: iOS/Collection View Management /Old/CollectionViewController.swift [hunk 3]
```diff
@@ -8,8 +8,8 @@
 
 import Foundation
 
-class CollectionViewController<SectionType: ManagerSectionType,
-                               ManagerType: CollectionViewManager<SectionType>>: ViewController {
+class CollectionViewController<SectionType: old_ManagerSectionType,
+                               ManagerType: old_CollectionViewManager<SectionType>>: ViewController {
```
Comment (MartinJibber) [inline iOS/Collection View Management /Old/CollectionViewController.swift:12]: Deprecating the old manager.

File: iOS/Collection View Management /UICollectionView.CellRegistration+Extensions.swift [hunk 4]
```diff
@@ -1,25 +0,0 @@
...
-//  UICollectionView.CellRegistration+Extensions.swift
-//  Ours
-//
-//  Created by Benji Dodgson on 2/12/21.
-//  Copyright © 2021 Benjamin Dodgson. All rights reserved.
-//
-
-import Foundation
-
-extension UICollectionView.CellRegistration where Cell: CollectionViewManagerCell & ManageableCell {
```
Comment (MartinJibber) [inline iOS/Collection View Management /UICollectionView.CellRegistration+Extensions.swift:11]: This wasn't being used.

File: iOS/Collection View Management /CollectionViewDataSourceCreator.swift [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+//  Copyright © 2021 Benjamin Dodgson. All rights reserved.
+//
+
+import UIKit
+
+/// A protocol for types that can create a UICollectionViewDiffableDataSource.
+/// Implementing types provide a means to dequeue collection view cells and supplementary views.
+/// A data source creation function is automatically implemented by default and uses the dequeing functions
+/// to initialize a new datasource.
+protocol CollectionViewDataSourceCreator {
```
Comment (MartinJibber) [inline iOS/Collection View Management /CollectionViewDataSourceCreator.swift:15]: Implementing types are responsible for creating diffable datasources and defining CellProvider functions. Importantly, DataSourceCreators DON'T hold onto any state. They don't hold a reference to the DiffableDataSource. They only create it.

I realized that the DiffableDataSource holds all the state we need and it only confuses things to wrap it up in another object. Instead we can interact with the DiffableDataSource directly and add convenience extensions if we find it awkward to manipulate the data.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**